af/analysis/ooni_db_backup.py

```diff
@@ -0,0 +1,287 @@
+from clickhouse_driver import Client as Clickhouse
+
+metrics = setup_metrics(name="db-backup")
+log = logging.getLogger("ooni-db-backup")
+
+# TODO backup schema
+
+# backup implemented in release 22.10, 2022-10-25
+sql_base = """
+    BACKUP TABLE %(tblname)s TO
+    S3(%(s3path_base)s, %(aws_id)s, %(aws_secret)s)
+    SETTINGS compression_method='lzma', compression_level=3
+"""
+sql_incremental = """
+    BACKUP TABLE %(tblname)s TO
+    S3(%(s3path_incremental)s, %(aws_id)s, %(aws_secret)s)
+    SETTINGS
+    base_backup = S3(%(s3path_base)s, %(aws_id)s, %(aws_secret)s)
+    compression_method='lzma', compression_level=3
+"""
+
+
+@metrics.timer("run_backup")
+def run_backup(conf) -> None:
+    aws_id = conf["public_aws_access_key_id"]
+    aws_secret = conf["public_aws_secret_access_key"]
+    bucket_name = conf["public_bucket_name"]
+    baseurl = f"https://{bucket_name}.s3.amazonaws.com/clickhouse_backup/"
+    click = Clickhouse.from_url(conf["clickhouse_url"])
+
+    for tblname, tbl_setting in conf["tables"].items():
+        if tbl_setting == "ignore":
+            log.info(f"Skipping {tblname}")
+            continue
+        elif tbl_setting == "full":
+            sql = sql_base
+        elif tbl_setting == "incremental":
+            sql = sql_incremental
+        else:
+            log.error(f"Skipping {tblname} unknown value {tbl_setting}")
+            continue
+
+        log.info(f"backing up {tblname}")
+        t0 = perf_counter()
+
+        s3path_base = f"{baseurl}/{tblname}__base.bkp"
+        s3path_incremental = f"{baseurl}/{tblname}__incremental.bkp"
+
+        kw = dict(
+            aws_id=aws_id,
+            aws_secret=aws_secret,
+            s3path_base=s3path_base,
+            s3path_incremental=s3path_incremental,
+            tblname=tblname,
+        )
+        click.execute(sql, kw)
+
+        delta_ms = int((perf_counter() - t0) * 1000)
+        metrics.timing(f"table_{tblname}", delta_ms)
+        log.info("table backup completed")
+
+
+def describe_table(click, tblname):
+    assert " " not in tblname
+    assert "'" not in tblname
+    colnames = []
+    blob = []
+    for row in click.execute(f"DESCRIBE TABLE {tblname}"):
+        colname, coltype = row[:2]
+        colnames.append(colname)
+        blob.append(f"{colname} {coltype}")
+
+    assert colnames
+    s_blob = ", ".join(blob)  # "measurement_uid String report_id String, ... "
+    s_colnames = ", ".join(colnames)  # "measurement_uid, report_id, ... "
+    return s_colnames, s_blob
+
+
+def export_fastpath(click, aws_id: str, aws_secret: str, baseurl: str) -> None:
+    tblname = "fastpath"
+    s_colnames, s_blob = describe_table(click, tblname)
+    assert s_blob.startswith("measurement_uid String, ")
+    # s_colnames is a string to be placed into ##colnames## without quotes
+    basesql = """
+    INSERT INTO FUNCTION s3(%(s3path)s, %(aws_id)s, %(aws_secret)s,
+        'CSVWithNames',
+        %(s_blob)s
+    )
+    SELECT ##colnames##
```

Comment:
What is the reason for expliciting the column names instead of just passing in `*`?

Using `*` seems more robust to potential schema changes.

Reply:
It is required by the current version of Clickhouse